In [1]:
from utils import css_from_file
css_from_file('style/style.css')

# Using Pandas read_json function:

In [91]:
import pandas as pd
import numpy as np

In [3]:
with open('yelp_dataset/yelp_academic_dataset_tip.json') as file:
        tip = pd.read_json(file, lines = True)

In [4]:
with open('yelp_dataset/yelp_academic_dataset_review.json') as file:
        review = pd.read_json(file, lines = True)

In [5]:
with open('yelp_dataset/yelp_academic_dataset_business.json') as file:
        business = pd.read_json(file, lines = True)

In [ ]:
with open('yelp_dataset/yelp_academic_dataset_checkin.json') as file:
        checkin = pd.read_json(file, lines = True)

In [6]:
with open('yelp_dataset/yelp_academic_dataset_user.json') as file:
        user = pd.read_json(file, lines = True)

In [8]:
# dfs = ['tip', 'review', 'business', 'checkin', 'user']

tip:  ['business_id' 'date' 'likes' 'text' 'type' 'user_id']
review:  ['business_id' 'date' 'review_id' 'stars' 'text' 'type' 'user_id' 'votes']
business:  ['attributes' 'business_id' 'categories' 'city' 'full_address' 'hours'
 'latitude' 'longitude' 'name' 'neighborhoods' 'open' 'review_count'
 'stars' 'state' 'type']
checkin:  ['business_id' 'checkin_info' 'type']
user:  ['average_stars' 'compliments' 'elite' 'fans' 'friends' 'name'
 'review_count' 'type' 'user_id' 'votes' 'yelping_since']

In [428]:
a = business.groupby(business.city)

In [433]:
pd.DataFrame({'count':a.size(), 'avg_stars': a.stars.mean()}).reset_index()

city  avg_stars  count
0                                  4.500000      2
1                  110 Las Vegas   5.000000      1
2                     Ahwahtukee   5.000000      1
3                      Ahwatukee   3.750000     12
4    Ahwatukee Foothills Village   5.000000      1
5                      Allegheny   5.000000      1
6                      Allentown   3.500000      2
7                          Anjou   3.318182     11
8                         Anthem   3.696296    135
9                Apache Junction   3.596774    186
10                     Arlington   3.750000      2
11             Arlington Heights   4.500000      1
12                     Asheville   4.500000      1
13                     Aspinwall   3.916667     12
14                        Avalon   3.500000      4
15                      Avondale   3.532787    427
16                   Baie-D'urfe   3.500000      1
17                       Balerno   2.000000      1
18                    Banksville   3.000000      1
19                  Beaconsfield   4.200000      5
20                      Bellevue   3.833333     24
21                       Bellvue   3.833333      3
22                       Belmont   3.759036     83
23                   Bethel Park   2.666667      3
24                    Bietigheim   3.625000      4
25             Black Canyon City   3.333333      3
26                    Blainville   3.750000      2
27                       Blawnox   4.250000      2
28                    Bloomfield   4.000000      1
29                  Blue Diamond   4.500000      1
..                           ...        ...    ...
414                  W Summerlin   3.000000      1
415                      Waddell   3.375000      8
416                    Waldbronn   4.166667      9
417               Water of Leith   4.500000      1
418                     Waterloo   3.496633    297
419                     Waunakee   3.597561     41
420                       Waxhaw   4.000000      5
421                   Weddington   4.500000      4
422                   Weingarten   4.090909     11
423           Weingarten (Baden)   4.166667      3
424                Wesley Chapel   3.625000      8
425               West Homestead   3.458333     12
426               West Las Vegas   2.500000      1
427                 West Mifflin   3.666667     36
428                    Westmount   3.745763     59
429         Westworld Scottsdale   3.000000      1
430                    Whitehall   3.000000      1
431                      Whitney   4.500000      1
432                   Wickenburg   3.794872     78
433                  Wilkinsburg   3.444444      9
434                      Windsor   3.400000     10
435                     Wittmann   3.500000      1
436                        Wörth   4.000000      1
437               Wörth am Rhein   4.062500      8
438                    Youngtown   4.000000     30
439                     glendale   4.333333      3
440                      phoenix   3.875000      4
441               pleasant hills   2.000000      1
442                   scottsdale   4.666667      3
443               Île des Soeurs   4.000000      1

[444 rows x 3 columns]

# Building a recommender system
First, get the relevant data: user-business stars... create matrices Y, R

** Y ** that contains #business as rows, #user as columns, values = stars

** R ** same size as Y, values = 1 if a user has rated a business **



## Need to determine businesses and users to include in the model

In [379]:
len(review.business_id.unique())

85539

In [380]:
len(review.user_id.unique())

686556

In [382]:
np.zeros([85539,686556])

MemoryError: 

## city-based grouping and recommendation

In [8]:
rev_biz = review.merge(business, right_on='business_id', left_on='business_id')

In [9]:
rev_biz.columns

Index(['business_id', 'date', 'review_id', 'stars_x', 'text', 'type_x',
       'user_id', 'votes', 'attributes', 'categories', 'city', 'full_address',
       'hours', 'latitude', 'longitude', 'name', 'neighborhoods', 'open',
       'review_count', 'stars_y', 'state', 'type_y'],
      dtype='object')

In [10]:
rev_biz.shape

(2685066, 22)

In [11]:
rev_biz_usr = rev_biz.merge(user, left_on = 'user_id', right_on='user_id')

In [12]:
rev_biz_usr.shape

(2685066, 32)

In [13]:
col_drop = rev_biz_usr.columns.difference(['business_id', 'user_id', 'stars_x'])
col_drop

Index(['attributes', 'average_stars', 'categories', 'city', 'compliments',
       'date', 'elite', 'fans', 'friends', 'full_address', 'hours', 'latitude',
       'longitude', 'name_x', 'name_y', 'neighborhoods', 'open',
       'review_count_x', 'review_count_y', 'review_id', 'stars_y', 'state',
       'text', 'type', 'type_x', 'type_y', 'votes_x', 'votes_y',
       'yelping_since'],
      dtype='object')

In [14]:
rev_biz_usr.drop(col_drop, axis=1, inplace=True)

In [16]:
rev_biz_usr.columns

Index(['business_id', 'stars_x', 'user_id'], dtype='object')

In [767]:
len(business.business_id.unique())

85900

In [17]:
len(rev_biz_usr.business_id.unique())

85539

In [768]:
len(user.user_id.unique())

686556

In [18]:
len(rev_biz_usr.user_id.unique())

686556

## The following pivot table gives an error due to out of memory, not negagtive dimensions:
memory needed assuming 8 bytes per value is:
686556 x 85539 x 8 / 1024 / 1024/ 1024 = 438 GB

In [19]:
686556 * 85539 * 8 / 1024 / 1024/ 1024

437.5525838434696

In [20]:
# rev_biz_usr.pivot(index='business_id', columns='user_id', values='stars_x')

In [21]:
# Y = pd.pivot_table(rev_biz_usr, index='business_id', columns='user_id', values='stars_x', dropna=False, aggfunc='mean')

In [23]:
from geopy.distance import vincenty

In [22]:
# vincenty((41.49008, -71.312796), (41.499498, -81.695391)).km

In [26]:
vincenty((80, -70), (80, -71)).km

19.39324680137632

In [27]:
vincenty((80, -71), (80, -72)).km

19.39324680137632

In [771]:
import shapely

In [7]:
import geopandas

# Check if cities and distances between them may be used instead of distances between businesses

##  No: cities are small, with distances that can be 3 miles

# assume 30 km as maximum distance willing to travel, to find vicinity within state

In [40]:
business.latitude.min(), business.latitude.max()

(32.865882499999998, 56.033777299999997)

In [59]:
business.longitude.min(), business.longitude.max()

(-115.38655, 8.5492494004572208)

In [41]:
vincenty((57, -71), (57, -72)).km

60.77164272792849

In [42]:
vincenty((32, -71), (32, -72)).km

94.49280502196765

In [58]:
vincenty((45, -71), (45, -72)).km

78.8463347093115

for out dataset, businesses lie within latitudes 32 and 56 North

where 1 degree in longitude ranges between 60 to 95 km

so nearby business ( < 30) should also be within 0.5 degree of longitude (near 32N distances will be < 47km)

for latitudes, one degree difference is about 110 km, so will adopt 0.3 difference to consider nearby businesses

In [53]:
vincenty((44, -71), (45, -71)).km

111.12200826241184

In [54]:
vincenty((44, -70), (45, -70)).km

111.12200826241184

In [55]:
vincenty((55, -71), (56, -71)).km

111.33269933263972

In [56]:
vincenty((32, -71), (33, -71)).km

110.89560891986069

In [65]:
business[business.latitude > 56.033]

attributes             business_id  \
47693         {}  4Yh57FsAyvDtZsq5F7fCWg   

                                              categories           city  \
47693  [Hotels & Travel, Public Transportation, Trans...  Inverkeithing   

                              full_address hours   latitude  longitude  \
47693  Boreland Rd\nInverkeithing KY11 1NJ    {}  56.033777  -3.393267   

                                name neighborhoods  open  review_count  stars  \
47693  Inverkeithing Railway Station            []  True             3    3.0   

      state      type  
47693   FIF  business

In [66]:
business.groupby(['state']).filter(lambda r: len(r) > 20).state.unique()

array(['PA', 'NC', 'SC', 'WI', 'IL', 'AZ', 'NV', 'QC', 'ON', 'EDH', 'MLN',
       'BW'], dtype=object)

In [67]:
# ignore states with less than 20 businesses... we have 12 states with more than 20 businesses
business_state = business.groupby(['state']).filter(lambda r: len(r) >= 20)

In [121]:
business_state.groupby('state').size()

state
AZ     36500
BW      1055
EDH     3297
IL       808
MLN      161
NC      6835
NV     23591
ON       530
PA      4086
QC      5591
SC       325
WI      3066
dtype: int64

In [127]:
len(business_state[business.state == 'AZ'].merge(review, how='inner', left_on='business_id', right_on='business_id').user_id.unique())

C:\Users\ali_a\Anaconda3\lib\site-packages\ipykernel\__main__.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if __name__ == '__main__':


247028

In [168]:
# To classify a business as Food:
business.attributes.iloc[3]['Good For']

{'breakfast': False,
 'brunch': False,
 'dessert': False,
 'dinner': False,
 'latenight': False,
 'lunch': True}

In [166]:
sum(business.attributes.iloc[3]['Good For'].values())

1

In [76]:
business_state.longitude.max()

8.5492494004572208

In [70]:
business_groups = business_state.groupby('state')

In [71]:
group = business_groups.get_group('EDH')

In [72]:
group.latitude.max()

56.000421000000003

In [766]:
nearby ={}
max_nearbies = 0

business_groups = business_state.groupby('state')

for name, group in business_groups:
    for i in range(0, len(group)):
        if group.latitude.iloc[i] < 80:
            near_group = group[(group.latitude < (group.latitude.iloc[i] + 1)) & (group.latitude > (group.latitude.iloc[i] - 1)) & 
                          (group.longitude < (group.longitude.iloc[i] + 1)) & (group.longitude > (group.longitude.iloc[i] - 1))]
        for j in range(0, len(near_group)):

            distance = vincenty(tuple(group.iloc[i][['latitude','longitude']]), tuple(near_group.iloc[j][['latitude','longitude']])).km
            if distance < 30:
                c += 1
                try:
                    nearby[group.business_id.iloc[i]].extend([near_group.business_id.iloc[j]])
                except KeyError as e:
                    nearby[group.business_id.iloc[i]] = [near_group.business_id.iloc[j]]
                try:
                    nearby[near_group.business_id.iloc[j]].extend([group.business_id.iloc[i]])
                except KeyError as e:
                    nearby[near_group.business_id.iloc[j]] = [group.business_id.iloc[i]]
        if max_nearbies < c:
            max_nearbies = c
    print (len(group))


KeyboardInterrupt: 

In [755]:
business_state[(abs(business_state.latitude) < (business_state.latitude[0] + 1))]

attributes  \
0      {'Take-out': True, 'Has TV': False, 'Noise Lev...   
1      {'Outdoor Seating': False, 'Good For Groups': ...   
2                                {'Good for Kids': True}   
3      {'Good For Groups': True, 'Has TV': True, 'Smo...   
4      {'Parking': {'valet': False, 'validated': Fals...   
5      {'Good For Groups': True, 'Has TV': True, 'Smo...   
6                         {'Accepts Credit Cards': True}   
7                                {'Good for Kids': True}   
8                                                     {}   
9                                                     {}   
10                                     {'Wi-Fi': 'free'}   
11                                                    {}   
12     {'Take-out': True, 'Good For Groups': True, 'P...   
13     {'Parking': {'valet': False, 'validated': Fals...   
14                                                    {}   
15                        {'Accepts Credit Cards': True}   
16                                                    {}   
17     {'Dogs Allowed': False, 'Good For Groups': Tru...   
18     {'Good For Groups': False, 'Has TV': True, 'Co...   
19     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
20     {'Price Range': 2, 'Wi-Fi': 'free', 'Accepts C...   
21     {'Take-out': True, 'Has TV': True, 'Noise Leve...   
22                                                    {}   
23                       {'Accepts Credit Cards': False}   
24     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
25                         {'By Appointment Only': True}   
26     {'Take-out': True, 'Has TV': True, 'Noise Leve...   
27                                                    {}   
28     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
29     {'Parking': {'valet': False, 'validated': Fals...   
...                                                  ...   
85866  {'Accepts Credit Cards': True, 'Good for Kids'...   
85867  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85868  {'Parking': {'valet': False, 'validated': Fals...   
85869  {'Outdoor Seating': True, 'Has TV': True, 'Whe...   
85870  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85871  {'Parking': {'valet': False, 'validated': Fals...   
85873                                                 {}   
85874  {'Outdoor Seating': True, 'Take-out': True, 'G...   
85875                                                 {}   
85878                            {'Good for Kids': True}   
85879                                                 {}   
85880  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85881  {'Price Range': 3, 'Take-out': True, 'Has TV':...   
85882  {'Parking': {'valet': False, 'validated': Fals...   
85883                                                 {}   
85884  {'Price Range': 1, 'Take-out': True, 'Has TV':...   
85885  {'Outdoor Seating': False, 'Take-out': True, '...   
85886  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85887  {'Parking': {'valet': False, 'validated': Fals...   
85888                     {'Accepts Credit Cards': True}   
85889  {'Dogs Allowed': True, 'Take-out': True, 'Has ...   
85890  {'Price Range': 2, 'Take-out': True, 'Has TV':...   
85891                                                 {}   
85892                     {'Accepts Credit Cards': True}   
85894  {'Parking': {'valet': False, 'validated': Fals...   
85895                                                 {}   
85896                     {'Accepts Credit Cards': True}   
85897  {'Take-out': True, 'Has TV': True, 'Noise Leve...   
85898                                                 {}   
85900                                                 {}   

                  business_id  \
0      5UmKMjUEUNdYWqANhGckJw   
1      UsFtqoBl7naz8AVUBZMjQQ   
2      cE27W9VPgO88Qxe4ol6y_g   
3      mVHrayjG3uZ_RLHkLj-AMg   
4      mYSpR_SLPgUVymYOvTQd_Q   
5      KayYbHCt-RkbGcPdGOThNg   
6      b12U9TFESStdy7CsTtcOeg   
7      Sktj1eHQFuVa-M4bgnEh8g   
8      3ZVKmuK2l7uXPE6lXY4Db

In [ ]:
len(nearby)

In [ ]:
import pickle
# save the nearby dictionary as it take a long time to build
with open('nearby_businesses_dict.pickle', 'wb') as handle:
    pickle.dump(nearby, handle, protocol=pickle.HIGHEST_PROTOCOL)
# To load the dictionary un-comment the following:___
# with open('nearby_businesses_dict.pickle', 'rb') as handle:
#     b = pickle.load(handle)

In [534]:
# nearby ={}
# max_nearbies = 0
# business_groups = business_state.groupby('state')
# for group in business_groups:
    
    
    
# for i in range(0, len(business)):
#     nearby[business.business_id.iloc[i]]=[]
#     c = 0
#     for j in range(0, len(business)):
#         if i==j: continue
#         distance = vincenty(tuple(business.iloc[i][['latitude','longitude']]), tuple(business.iloc[j][['latitude','longitude']])).km
#         if distance < 30:
#             c += 1
#             nearby[business.business_id.iloc[i]].extend([business.business_id.iloc[j]])        
#     if c > max_nearbies:
#         max_nearbies = c

In [170]:
pd.__version__

'0.19.2'